lib/indexer/tag_lookup.rb

```diff
@@ -11,43 +13,54 @@ def prepare_tags(doc_hash)
       # field). These won't have tags associated with them so we can bail out.
       return doc_hash if doc_hash["link"] =~ /\Ahttps?:\/\//
 
-      artefact = find_document_from_content_api(doc_hash["link"])
-      return doc_hash unless artefact
-      add_tags_from_artefact(artefact, doc_hash)
+      content_id = find_content_id(doc_hash)
+      return doc_hash unless content_id
+
+      links = find_links(content_id)
+      return doc_hash unless links
+
+      merge_links_into_payload(links, doc_hash)
     end
 
   private
 
-    def find_document_from_content_api(link)
-      link_without_trailing_slash = link.sub(/\A\//, '')
-      begin
-        Services.content_api.artefact!(link_without_trailing_slash)
-      rescue GdsApi::HTTPNotFound, GdsApi::HTTPGone
-        nil
+    # Some applications send the `content_id` for their items. This means we can
+    # skip the lookup from the publishing-api.
+    def find_content_id(doc_hash)
+      if doc_hash["content_id"].present?
+        doc_hash["content_id"]
+      else
+        Services.publishing_api.lookup_content_id(base_path: doc_hash["link"])
       end
     end
 
-    def add_tags_from_artefact(artefact, doc_hash)
-      doc_hash["organisations"] ||= []
-      doc_hash["mainstream_browse_pages"] ||= []
-      doc_hash["specialist_sectors"] ||= []
-
-      artefact.tags.each do |tag|
-        case tag.details.type
-        when "organisation"
-          doc_hash["organisations"] << tag.slug
-        when "section"
-          doc_hash["mainstream_browse_pages"] << tag.slug
-        when "specialist_sector"
-          doc_hash["specialist_sectors"] << tag.slug
-        end
-      end
-
-      doc_hash["organisations"].uniq!
-      doc_hash["mainstream_browse_pages"].uniq!
-      doc_hash["specialist_sectors"].uniq!
+    def find_links(content_id)
+      Services.publishing_api.get_links(content_id).links
+    end
 
+    def merge_links_into_payload(links, doc_hash)
+      doc_hash["specialist_sectors"] = content_ids_to_slugs(links["topics"].to_a)
+      doc_hash["mainstream_browse_pages"] = content_ids_to_slugs(links["mainstream_browse_pages"].to_a)
+      doc_hash["organisations"] = content_ids_to_slugs(links["organisations"].to_a)
       doc_hash
     end
+
+    # This lookup works because each link type in `links` contains items with
+    # format specialist sector, mainstream browse page or organisation. We have
+    # ensured that they have a content_id by updating whitehall &
+    # collections-publisher accordingly.
+    def content_ids_to_slugs(content_ids)
+      content_ids.map do |content_id|
+        result = client.raw_search(query: { term: { content_id: content_id } }, size: 1)
```

Comment:
Chatty vs non-chatty - common but not ubiquitous phrase to describe the granularity of api requests available. A chatty interface is one which provides very small, modular functionality, so is usually more flexible but requires more calls to get something useful done. A non-chatty interface provides fewer, higher level calls which are enough to do something useful individually.
The design tradeoff involves complexity of client, complexity of service, flexibility, and invocation overhead.
Here, aiui, we use
- one api call to get the content id for a slug
- one api call to get the links for a content id
- some undetermined mechanism to get the slugs for the link ids
- all per-content-item

The minimal non-chatty interface would be
- one batch call to get the content id, link ids and link slugs for several item slugs simultaneously

But of course there is a design continuum. Knowing nothing about the api in question, my first suggestion would be to add the link slugs to the payload returned from the get_links lookup.